File: internal/channel/discord.go
package channel

import (
	"fmt"
	"log/slog"
	"strings"
	"sync"
	"time"

	"github.com/Z-M-Huang/openhive/internal/domain"
	"github.com/bwmarrin/discordgo"
)

const (
	discordPrefix       = "discord"
	discordMaxChunkSize = 2000
	discordRateWindow   = 5 * time.Second
	discordRateLimit    = 5
)

// DiscordSession is an interface wrapping the discordgo.Session methods we need,
// allowing injection of a mock in tests.
type DiscordSession interface {
	Open() error
	Close() error
	ChannelMessageSend(channelID string, content string, options ...discordgo.RequestOption) (*discordgo.Message, error)
	AddHandler(handler interface{}) func()
	State() *discordgo.State
}

// discordgoSessionAdapter wraps a *discordgo.Session to satisfy DiscordSession.
type discordgoSessionAdapter struct {
	s *discordgo.Session
}

func (a *discordgoSessionAdapter) Open() error  { return a.s.Open() }
func (a *discordgoSessionAdapter) Close() error { return a.s.Close() }
func (a *discordgoSessionAdapter) ChannelMessageSend(channelID string, content string, options ...discordgo.RequestOption) (*discordgo.Message, error) {
	return a.s.ChannelMessageSend(channelID, content, options...)
}
func (a *discordgoSessionAdapter) AddHandler(handler interface{}) func() {
	return a.s.AddHandler(handler)
}
func (a *discordgoSessionAdapter) State() *discordgo.State {
	return a.s.State
}

// DiscordSessionFactory creates a DiscordSession from a token.
// Overridable in tests.
type DiscordSessionFactory func(token string) (DiscordSession, error)

// defaultDiscordFactory creates a real discordgo session.
func defaultDiscordFactory(token string) (DiscordSession, error) {
	s, err := discordgo.New("Bot " + token)
	if err != nil {
		return nil, fmt.Errorf("failed to create discord session: %w", err)
	}
	s.Identify.Intents = discordgo.IntentsGuildMessages | discordgo.IntentsDirectMessages
	return &discordgoSessionAdapter{s: s}, nil
}

// DiscordChannel implements domain.ChannelAdapter for Discord.
type DiscordChannel struct {
	session    DiscordSession
	channelID  string
	token      string
	enabled    bool
	factory    DiscordSessionFactory
	eventBus   domain.EventBus
	logger     *slog.Logger
	onMessage  func(jid string, content string)
	onMetadata func(jid string, metadata map[string]string)
	mu         sync.RWMutex
	connected  bool

	// botUserID is set after Connect() resolves the bot's own user ID
	botUserID string

	// Rate limiting: track per-channel send times
	rateMu    sync.Mutex
	sendTimes []time.Time
}

// DiscordConfig holds the configuration for the Discord channel.
type DiscordConfig struct {
	Token     string
	ChannelID string
	Enabled   bool
}

// NewDiscordChannel creates a new Discord channel adapter.
// factory may be nil to use the default discordgo factory.
func NewDiscordChannel(cfg DiscordConfig, eventBus domain.EventBus, logger *slog.Logger, factory DiscordSessionFactory) *DiscordChannel {
	if factory == nil {
		factory = defaultDiscordFactory
	}
	return &DiscordChannel{
		token:     cfg.Token,
		channelID: cfg.ChannelID,
		enabled:   cfg.Enabled,
		factory:   factory,
		eventBus:  eventBus,
		logger:    logger,
	}
}

// HandleConfigChange reacts to a MasterConfig change event.
// It compares the new Discord channel config to the current state and
// connects, disconnects, or reconnects as needed.
//
// The caller is responsible for passing decrypted tokens — this method
// does not decrypt enc:-prefixed values.
func (d *DiscordChannel) HandleConfigChange(newToken, newChannelID string, newEnabled bool) {
	d.mu.Lock()
	oldEnabled := d.enabled
	oldToken := d.token
	oldChannelID := d.channelID
	wasConnected := d.connected
	d.mu.Unlock()

	switch {
	case oldEnabled && !newEnabled:
		// Was enabled, now disabled: disconnect.
		d.logger.Info("discord: config changed — disabling channel")
		if err := d.Disconnect(); err != nil {
			d.logger.Warn("discord: disconnect on disable failed", "error", err)
		}
		d.mu.Lock()
		d.enabled = false
		d.mu.Unlock()

	case !oldEnabled && newEnabled:
		// Was disabled, now enabled: update credentials and connect.
		d.logger.Info("discord: config changed — enabling channel")
		d.mu.Lock()
		d.token = newToken
		d.channelID = newChannelID
		d.enabled = true
		d.mu.Unlock()
		if err := d.Connect(); err != nil {
			d.logger.Warn("discord: connect on enable failed", "error", err)
		}

	case newEnabled && wasConnected && (newToken != oldToken || newChannelID != oldChannelID):
		// Credentials changed while connected: reconnect with new credentials.
		d.logger.Info("discord: config changed — reconnecting with new credentials")
		if err := d.Disconnect(); err != nil {
			d.logger.Warn("discord: disconnect before credential update failed", "error", err)
		}
		d.mu.Lock()
		d.token = newToken
		d.channelID = newChannelID
		d.enabled = true
		d.mu.Unlock()
		if err := d.Connect(); err != nil {
			d.logger.Warn("discord: reconnect after credential update failed", "error", err)
		}

	default:
		// No actionable change (e.g. still disabled, or enabled+connected with same credentials).
		d.mu.Lock()
		d.token = newToken
		d.channelID = newChannelID
		d.enabled = newEnabled
		d.mu.Unlock()
	}
}

// Connect opens the Discord session and registers the message handler.
func (d *DiscordChannel) Connect() error {
	d.mu.Lock()
	defer d.mu.Unlock()

	if d.connected {
		return nil
	}

	s, err := d.factory(d.token)
	if err != nil {
		return fmt.Errorf("discord connect: failed to create session: %w", err)
	}

	d.session = s

	// Register the message-create handler before opening the session.
	s.AddHandler(d.handleMessageCreate)

	if err := s.Open(); err != nil {
		return fmt.Errorf("discord connect: failed to open session: %w", err)
	}

	// Retrieve the bot's own user ID from the session state (via embedded Ready.User).
	if state := s.State(); state != nil && state.Ready.User != nil {
		d.botUserID = state.Ready.User.ID
	}

	d.connected = true
	d.logger.Info("discord channel connected", "channel_id", d.channelID)
	return nil
}

// Disconnect closes the Discord session.
func (d *DiscordChannel) Disconnect() error {
	d.mu.Lock()
	defer d.mu.Unlock()

	if !d.connected {
		return nil
	}
	if d.session != nil {
		if err := d.session.Close(); err != nil {
			d.logger.Warn("discord disconnect: session close error", "error", err)
		}
	}
	d.connected = false
	d.logger.Info("discord channel disconnected", "channel_id", d.channelID)
	return nil
}

// SendMessage sends a message to the Discord channel, splitting long content
// at paragraph or sentence boundaries so each chunk is <=2000 characters.
// Rate-limited to 5 messages per 5 seconds.
func (d *DiscordChannel) SendMessage(jid string, content string) error {
	d.mu.RLock()
	sess := d.session
	connected := d.connected
	d.mu.RUnlock()

	if !connected || sess == nil {
		return &domain.ValidationError{Field: "connection", Message: "discord channel is not connected"}
	}

	chunks := splitMessage(content, discordMaxChunkSize)
	for _, chunk := range chunks {
		if err := d.sendWithRateLimit(sess, chunk); err != nil {
			return err
		}
	}
	return nil
}

// GetJIDPrefix returns the JID prefix for Discord.
func (d *DiscordChannel) GetJIDPrefix() string {
	return discordPrefix
}

// IsConnected returns the current connection state.
func (d *DiscordChannel) IsConnected() bool {
	d.mu.RLock()
	defer d.mu.RUnlock()
	return d.connected
}

// OnMessage sets the callback for incoming messages.
func (d *DiscordChannel) OnMessage(callback func(jid string, content string)) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.onMessage = callback
}

// OnMetadata sets the callback for metadata events (unused by Discord).
func (d *DiscordChannel) OnMetadata(callback func(jid string, metadata map[string]string)) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.onMetadata = callback
}

// handleMessageCreate is the discordgo message-create event handler.
// Filters out bot/webhook/wrong-channel/self messages, then calls onMessage.
func (d *DiscordChannel) handleMessageCreate(s *discordgo.Session, m *discordgo.MessageCreate) {
	// Filter: nil message or author
	if m == nil || m.Message == nil || m.Author == nil {
		return
	}

	// Filter: bot messages
	if m.Author.Bot {
		return
	}

	// Filter: webhook messages
	if m.WebhookID != "" {
		return
	}

	// Filter: wrong channel
	d.mu.RLock()
	channelID := d.channelID
	botUserID := d.botUserID
	d.mu.RUnlock()

	if channelID != "" && m.ChannelID != channelID {
		return
	}

	// Filter: self messages
	if botUserID != "" && m.Author.ID == botUserID {
		return
	}

	content := strings.TrimSpace(m.Content)
	if content == "" {
		return
	}

	jid := fmt.Sprintf("%s:%s:%s", discordPrefix, m.ChannelID, m.Author.ID)

	d.mu.RLock()
	handler := d.onMessage
	d.mu.RUnlock()

	if handler != nil {
		handler(jid, content)
	}
}

// sendWithRateLimit sends a single message chunk, enforcing the rate limit.
func (d *DiscordChannel) sendWithRateLimit(sess DiscordSession, chunk string) error {
	d.rateMu.Lock()
	now := time.Now()
	// Evict sends older than the rate window
	valid := d.sendTimes[:0]
	for _, t := range d.sendTimes {
		if now.Sub(t) < discordRateWindow {
			valid = append(valid, t)
		}
	}
	d.sendTimes = valid

	if len(d.sendTimes) >= discordRateLimit {
		d.rateMu.Unlock()
		return &domain.ValidationError{
			Field:   "rate_limit",
			Message: fmt.Sprintf("discord rate limit exceeded: max %d messages per %s", discordRateLimit, discordRateWindow),
		}
	}
	d.sendTimes = append(d.sendTimes, now)
	d.rateMu.Unlock()

	_, err := sess.ChannelMessageSend(d.channelID, chunk)
	if err != nil {
		return fmt.Errorf("discord send message: %w", err)
	}
	return nil
}

// HandleMessageCreateForTest exposes the internal message handler for integration tests.
// It should only be called in test code.
func (d *DiscordChannel) HandleMessageCreateForTest(s *discordgo.Session, m *discordgo.MessageCreate) {
	d.handleMessageCreate(s, m)
}

// splitMessage splits content into chunks of at most maxLen characters.
// It prefers to split at paragraph boundaries (\n\n), then sentence boundaries (. ! ?),
// then word boundaries, and as a last resort at the hard character limit.
func splitMessage(content string, maxLen int) []string {
	if len(content) <= maxLen {
		return []string{content}
	}

	var chunks []string
	remaining := content

	for len(remaining) > maxLen {
		chunk := remaining[:maxLen]

		// Try to split at a paragraph boundary (\n\n)
		if idx := strings.LastIndex(chunk, "\n\n"); idx > 0 {
			chunks = append(chunks, strings.TrimRight(remaining[:idx], " \t"))
			remaining = strings.TrimLeft(remaining[idx+2:], " \t")
			continue
		}

		// Try to split at a sentence boundary (. ! ?)
		splitIdx := -1
		for i := len(chunk) - 1; i >= 0; i-- {
			if chunk[i] == '.' || chunk[i] == '!' || chunk[i] == '?' {
				// Only split if followed by a space or newline (or end of chunk)
				if i+1 < len(chunk) && (chunk[i+1] == ' ' || chunk[i+1] == '\n') {
					splitIdx = i + 1
					break
				}
			}
		}
		if splitIdx > 0 {
			chunks = append(chunks, strings.TrimRight(remaining[:splitIdx], " \t"))
			remaining = strings.TrimLeft(remaining[splitIdx:], " \t")
			continue
		}

		// Try to split at a word boundary (space or newline)
		for i := len(chunk) - 1; i >= 0; i-- {
			if chunk[i] == ' ' || chunk[i] == '\n' {
				splitIdx = i
				break
			}
		}
		if splitIdx > 0 {
			chunks = append(chunks, strings.TrimRight(remaining[:splitIdx], " \t"))
			remaining = strings.TrimLeft(remaining[splitIdx:], " \t")
			continue
		}

		// Hard split at maxLen
		chunks = append(chunks, remaining[:maxLen])
		remaining = remaining[maxLen:]
	}

	if strings.TrimSpace(remaining) != "" {
		chunks = append(chunks, remaining)
	}
	return chunks
}

File: backend/src/channel/discord.ts
/**
 * OpenHive Backend - Discord Channel Adapter
 *
 * Implements ChannelAdapter for Discord using discord.js v14. Bot message
 * handling with filtering (bots, webhooks, wrong channel, self), rate limiting
 * (5 messages / 5 s), message chunking (2000-char limit with paragraph /
 * sentence / word boundary splitting), automatic reconnection with exponential
 * backoff, and config hot-reload.
 *
 * Implements automatic reconnection with exponential backoff and config hot-reload.
 */

import {
  Client,
  GatewayIntentBits,
  type Message,
  type OmitPartialGroupDMChannel,
  type TextChannel,
  type DMChannel,
  type NewsChannel,
} from 'discord.js';

import type { ChannelAdapter } from '../domain/interfaces.js';
import { ValidationError } from '../domain/errors.js';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const DISCORD_PREFIX = 'discord';
const MAX_CHUNK_SIZE = 2000;
const RATE_WINDOW_MS = 5000;
const RATE_LIMIT = 5;
const BACKOFF_BASE_MS = 1000;
const BACKOFF_MAX_MS = 60000;
const BACKOFF_MULTIPLIER = 2;

// ---------------------------------------------------------------------------
// DiscordConfig
// ---------------------------------------------------------------------------

/**
 * Configuration for the Discord channel adapter.
 */
export interface DiscordConfig {
  token: string;
  channelID: string;
  enabled: boolean;
}

// ---------------------------------------------------------------------------
// DiscordClientInterface
// ---------------------------------------------------------------------------

/**
 * Abstraction over discord.js Client methods needed by DiscordChannel.
 * Allows injection of a mock in tests without importing discord.js Client.
 *
 * The `on` method covers event registration. The `login` method connects
 * to Discord. `destroy` disconnects. `fetchChannel` wraps client.channels.fetch.
 */
export interface DiscordClientInterface {
  login(token: string): Promise<string>;
  destroy(): Promise<void>;
  fetchChannel(channelID: string): Promise<TextChannel | DMChannel | NewsChannel | null>;
  getBotUserID(): string | null;
  on(event: 'messageCreate', listener: (message: OmitPartialGroupDMChannel<Message>) => void): void;
  on(event: 'error', listener: (error: Error) => void): void;
  on(event: 'shardDisconnect', listener: (closeEvent: CloseEvent, shardId: number) => void): void;
  on(event: string, listener: (...args: unknown[]) => void): void;
}

// ---------------------------------------------------------------------------
// RealDiscordClient  — production wrapper over discord.js Client
// ---------------------------------------------------------------------------

/**
 * Wraps a real discord.js Client to satisfy DiscordClientInterface.
 * Created internally by DiscordChannel.connect() in production.
 */
class RealDiscordClient implements DiscordClientInterface {
  private readonly client: Client;

  constructor() {
    this.client = new Client({
      intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildMessages,
        GatewayIntentBits.DirectMessages,
        GatewayIntentBits.MessageContent,
      ],
    });
  }

  login(token: string): Promise<string> {
    return this.client.login(token);
  }

  async destroy(): Promise<void> {
    await this.client.destroy();
  }

  async fetchChannel(channelID: string): Promise<TextChannel | DMChannel | NewsChannel | null> {
    const channel = await this.client.channels.fetch(channelID);
    if (channel === null) return null;
    // Only accept text-based channels that support send()
    if (
      channel.isTextBased() &&
      (channel.type === 0 || // ChannelType.GuildText
        channel.type === 1 || // ChannelType.DM
        channel.type === 5) // ChannelType.GuildAnnouncement
    ) {
      return channel as TextChannel | DMChannel | NewsChannel;
    }
    return null;
  }

  getBotUserID(): string | null {
    return this.client.user?.id ?? null;
  }

  on(event: 'messageCreate', listener: (message: OmitPartialGroupDMChannel<Message>) => void): void;
  on(event: 'error', listener: (error: Error) => void): void;
  on(event: 'shardDisconnect', listener: (closeEvent: CloseEvent, shardId: number) => void): void;
  on(event: string, listener: (...args: unknown[]) => void): void;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  on(event: string, listener: (...args: any[]) => void): void {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (this.client as any).on(event, listener);
  }
}

// ---------------------------------------------------------------------------
// DiscordClientFactory
// ---------------------------------------------------------------------------

/**
 * Factory function type for creating DiscordClientInterface instances.
 * Injectable for test isolation.
 */
export type DiscordClientFactory = () => DiscordClientInterface;

/** Default production factory. */
const defaultClientFactory: DiscordClientFactory = () => new RealDiscordClient();

// ---------------------------------------------------------------------------
// SleepFn
// ---------------------------------------------------------------------------

/**
 * Async sleep function. Injectable for test isolation (tests pass an
 * immediate-resolve mock so no real delays occur in unit tests).
 */
export type SleepFn = (ms: number) => Promise<void>;

const defaultSleepFn: SleepFn = (ms: number): Promise<void> =>
  new Promise(resolve => setTimeout(resolve, ms));

// ---------------------------------------------------------------------------
// Logger interface
// ---------------------------------------------------------------------------

/**
 * Minimal logger subset used by DiscordChannel.
 */
export interface DiscordLogger {
  info(msg: string, ctx?: Record<string, unknown>): void;
  warn(msg: string, ctx?: Record<string, unknown>): void;
  error(msg: string, ctx?: Record<string, unknown>): void;
}

// ---------------------------------------------------------------------------
// DiscordChannelOptions
// ---------------------------------------------------------------------------

/**
 * Optional overrides for DiscordChannel — primarily for test injection.
 */
export interface DiscordChannelOptions {
  /** Overrides the discord.js Client factory. */
  clientFactory?: DiscordClientFactory;
  /**
   * Overrides the sleep function used in reconnectLoop().
   * Pass an immediately-resolving stub in tests to skip real delays.
   */
  sleepFn?: SleepFn;
}

// ---------------------------------------------------------------------------
// DiscordChannel
// ---------------------------------------------------------------------------

/**
 * Discord channel adapter implementing ChannelAdapter.
 *
 * Key behaviour:
 *  - connect() creates a Client, registers handlers, logs in, captures bot UID.
 *  - disconnect() sets shouldReconnect=false, destroys the client.
 *  - sendMessage() chunks content and sends each chunk with rate limiting.
 *  - handleMessageCreate() filters bots / webhooks / wrong channel / self.
 *  - reconnectLoop() is triggered by 'error' and 'shardDisconnect' events and
 *    retries with exponential backoff (backoffBase * 2^attempt, cap 60 s).
 *  - handleConfigChange() compares state and connect/disconnect/reconnects.
 */
export class DiscordChannel implements ChannelAdapter {
  // Config state
  private token: string;
  private channelID: string;
  private enabled: boolean;

  // Dependencies
  private readonly clientFactory: DiscordClientFactory;
  private readonly sleepFn: SleepFn;
  private readonly logger: DiscordLogger;

  // Runtime state
  private client: DiscordClientInterface | null = null;
  private connected: boolean = false;
  private botUserID: string | null = null;

  // Callbacks registered by the MessageRouter
  private messageCallback: ((jid: string, content: string) => void) | null = null;
  private metadataCallback: ((jid: string, metadata: Record<string, string>) => void) | null = null;

  // Reconnection state
  private shouldReconnect: boolean = false;
  private reconnectAttempt: number = 0;

  // Rate limiting: timestamps of recent sends (ms)
  private sendTimes: number[] = [];

  constructor(cfg: DiscordConfig, logger: DiscordLogger, options?: DiscordChannelOptions) {
    this.token = cfg.token;
    this.channelID = cfg.channelID;
    this.enabled = cfg.enabled;
    this.logger = logger;
    this.clientFactory = options?.clientFactory ?? defaultClientFactory;
    this.sleepFn = options?.sleepFn ?? defaultSleepFn;
  }

  // ---------------------------------------------------------------------------
  // ChannelAdapter — connect
  // ---------------------------------------------------------------------------

  /**
   * Creates a discord.js Client, registers event handlers, and logs in.
   * Resolves bot user ID after login. Sets shouldReconnect=true so the
   * reconnect loop will trigger on future 'error' / 'shardDisconnect' events.
   */
  async connect(): Promise<void> {
    if (this.connected) {
      return;
    }

    const client = this.clientFactory();
    this.client = client;
    this.shouldReconnect = true;

    // Register inbound message handler
    client.on(
      'messageCreate',
      (message: OmitPartialGroupDMChannel<Message>) => {
        this.handleMessageCreate(message);
      },
    );

    // Register reconnect triggers
    client.on('error', (_error: Error) => {
      if (this.shouldReconnect) {
        void this.reconnectLoop();
      }
    });

    client.on('shardDisconnect', (_closeEvent: CloseEvent, _shardId: number) => {
      if (this.shouldReconnect) {
        void this.reconnectLoop();
      }
    });

    await client.login(this.token);

    // Capture bot user ID (set by discord.js after ready)
    this.botUserID = client.getBotUserID();
    this.connected = true;
    this.reconnectAttempt = 0;

    this.logger.info('discord channel connected', { channel_id: this.channelID });
  }

  // ---------------------------------------------------------------------------
  // ChannelAdapter — disconnect
  // ---------------------------------------------------------------------------

  /**
   * Stops the reconnect loop and destroys the client.
   */
  async disconnect(): Promise<void> {
    this.shouldReconnect = false;

    if (!this.connected || this.client === null) {
      return;
    }

    try {
      await this.client.destroy();
    } catch (err) {
      this.logger.warn('discord disconnect: destroy error', {
        error: err instanceof Error ? err.message : String(err),
      });
    }

    this.client = null;
    this.connected = false;
    this.botUserID = null;
    this.logger.info('discord channel disconnected', { channel_id: this.channelID });
  }

  // ---------------------------------------------------------------------------
  // ChannelAdapter — sendMessage
  // ---------------------------------------------------------------------------

  /**
   * Splits content into chunks of at most 2000 chars and sends each with rate
   * limiting. Throws if not connected or rate limit exceeded.
   */
  async sendMessage(_jid: string, content: string): Promise<void> {
    if (!this.connected || this.client === null) {
      throw new ValidationError('connection', 'discord channel is not connected');
    }

    const channel = await this.client.fetchChannel(this.channelID);
    if (channel === null) {
      throw new ValidationError('channel', `discord channel not found: ${this.channelID}`);
    }

    const chunks = splitMessage(content, MAX_CHUNK_SIZE);
    for (const chunk of chunks) {
      this.sendWithRateLimit(chunk);
      await channel.send(chunk);
    }
  }

  // ---------------------------------------------------------------------------
  // ChannelAdapter — getJIDPrefix / isConnected / onMessage / onMetadata
  // ---------------------------------------------------------------------------

  getJIDPrefix(): string {
    return DISCORD_PREFIX;
  }

  isConnected(): boolean {
    return this.connected;
  }

  onMessage(callback: (jid: string, content: string) => void): void {
    this.messageCallback = callback;
  }

  onMetadata(callback: (jid: string, metadata: Record<string, string>) => void): void {
    this.metadataCallback = callback;
    void this.metadataCallback; // stored for future use; satisfies noUnusedLocals
  }

  // ---------------------------------------------------------------------------
  // handleConfigChange
  // ---------------------------------------------------------------------------

  /**
   * Compares new config to current state and connect / disconnect / reconnect
   * as needed. Resets reconnectAttempt on any manual reconnect.
   */
  async handleConfigChange(
    newToken: string,
    newChannelID: string,
    newEnabled: boolean,
  ): Promise<void> {
    const oldEnabled = this.enabled;
    const oldToken = this.token;
    const oldChannelID = this.channelID;
    const wasConnected = this.connected;

    if (oldEnabled && !newEnabled) {
      // Was enabled, now disabled: disconnect.
      this.logger.info('discord: config changed — disabling channel');
      this.enabled = false;
      await this.disconnect().catch((err: unknown) => {
        this.logger.warn('discord: disconnect on disable failed', {
          error: err instanceof Error ? err.message : String(err),
        });
      });
    } else if (!oldEnabled && newEnabled) {
      // Was disabled, now enabled: update credentials and connect.
      this.logger.info('discord: config changed — enabling channel');
      this.token = newToken;
      this.channelID = newChannelID;
      this.enabled = true;
      this.reconnectAttempt = 0;
      await this.connect().catch((err: unknown) => {
        this.logger.warn('discord: connect on enable failed', {
          error: err instanceof Error ? err.message : String(err),
        });
      });
    } else if (newEnabled && wasConnected && (newToken !== oldToken || newChannelID !== oldChannelID)) {
      // Credentials changed while connected: reconnect.
      this.logger.info('discord: config changed — reconnecting with new credentials');
      await this.disconnect().catch((err: unknown) => {
        this.logger.warn('discord: disconnect before credential update failed', {
          error: err instanceof Error ? err.message : String(err),
        });
      });
      this.token = newToken;
      this.channelID = newChannelID;
      this.enabled = true;
      this.reconnectAttempt = 0;
      await this.connect().catch((err: unknown) => {
        this.logger.warn('discord: reconnect after credential update failed', {
          error: err instanceof Error ? err.message : String(err),
        });
      });
    } else {
      // No actionable change — just update stored state.
      this.token = newToken;
      this.channelID = newChannelID;
      this.enabled = newEnabled;
    }
  }

  // ---------------------------------------------------------------------------
  // Private — handleMessageCreate
  // ---------------------------------------------------------------------------

  /**
   * Filters and dispatches an inbound discord.js message.
   * Filters: bots, webhooks, wrong channel, self, empty content.
   * Builds JID as "discord:<channelId>:<authorId>".
   */
  handleMessageCreate(message: OmitPartialGroupDMChannel<Message>): void {
    // Filter: bot messages
    if (message.author.bot) {
      return;
    }

    // Filter: webhook messages
    if (message.webhookId !== null) {
      return;
    }

    // Filter: wrong channel
    if (this.channelID !== '' && message.channelId !== this.channelID) {
      return;
    }

    // Filter: self messages
    if (this.botUserID !== null && message.author.id === this.botUserID) {
      return;
    }

    const content = message.content.trim();
    if (content === '') {
      return;
    }

    const jid = `${DISCORD_PREFIX}:${message.channelId}:${message.author.id}`;

    if (this.messageCallback !== null) {
      this.messageCallback(jid, content);
    }
  }

  // ---------------------------------------------------------------------------
  // Private — sendWithRateLimit
  // ---------------------------------------------------------------------------

  /**
   * Enforces the rate limit: at most RATE_LIMIT sends per RATE_WINDOW_MS.
   * Evicts old entries, then throws ValidationError if over limit.
   * Does NOT perform the actual send — the channel.send() call is in sendMessage().
   */
  private sendWithRateLimit(_chunk: string): void {
    const now = Date.now();
    // Evict entries outside the rate window
    this.sendTimes = this.sendTimes.filter(t => now - t < RATE_WINDOW_MS);

    if (this.sendTimes.length >= RATE_LIMIT) {
      throw new ValidationError(
        'rate_limit',
        `discord rate limit exceeded: max ${RATE_LIMIT} messages per ${RATE_WINDOW_MS}ms`,
      );
    }

    this.sendTimes.push(now);
  }

  // ---------------------------------------------------------------------------
  // reconnectLoop (exported as public for direct testing)
  // ---------------------------------------------------------------------------

  /**
   * Exponential backoff reconnection loop. Called when 'error' or
   * 'shardDisconnect' events fire. Retries connect() while shouldReconnect
   * is true. Resets reconnectAttempt on success.
   *
   * Delay = backoffBase * (backoffMultiplier ^ attempt), capped at backoffMax.
   *
   * Marked public so tests can call it directly and spy on it.
   */
  async reconnectLoop(): Promise<void> {
    // Mark disconnected so state is accurate during retry waits
    this.connected = false;
    this.client = null;

    while (this.shouldReconnect) {
      const delay = Math.min(
        BACKOFF_BASE_MS * Math.pow(BACKOFF_MULTIPLIER, this.reconnectAttempt),
        BACKOFF_MAX_MS,
      );

      this.logger.info('discord: reconnecting', {
        attempt: this.reconnectAttempt,
        delay_ms: delay,
      });

      await this.sleepFn(delay);

      if (!this.shouldReconnect) {
        break;
      }

      try {
        await this.connect();
        // Success — reset backoff
        this.reconnectAttempt = 0;
        this.logger.info('discord: reconnected successfully');
        return;
      } catch (err) {
        this.reconnectAttempt++;
        this.logger.warn('discord: reconnect attempt failed', {
          attempt: this.reconnectAttempt,
          error: err instanceof Error ? err.message : String(err),
        });
      }
    }
  }
}

// ---------------------------------------------------------------------------
// splitMessage — exported pure function for direct testing
// ---------------------------------------------------------------------------

/**
 * Splits content into chunks of at most maxLen characters.
 * Prefers paragraph (\n\n) > sentence (. ! ?) > word > hard split.
 *
 * Prefers paragraph (\n\n) > sentence (. ! ?) > word > hard split.
 */
export function splitMessage(content: string, maxLen: number): string[] {
  if (content.length <= maxLen) {
    return [content];
  }

  const chunks: string[] = [];
  let remaining = content;

  while (remaining.length > maxLen) {
    const chunk = remaining.slice(0, maxLen);

    // Prefer paragraph boundary (\n\n)
    const paraIdx = chunk.lastIndexOf('\n\n');
    if (paraIdx > 0) {
      chunks.push(remaining.slice(0, paraIdx).trimEnd());
      remaining = remaining.slice(paraIdx + 2).trimStart();
      continue;
    }

    // Prefer sentence boundary (. ! ?) followed by space or \n
    let sentIdx = -1;
    for (let i = chunk.length - 1; i >= 0; i--) {
      const ch = chunk[i];
      if (ch === '.' || ch === '!' || ch === '?') {
        const next = chunk[i + 1];
        if (next === ' ' || next === '\n' || next === undefined) {
          sentIdx = i + 1;
          break;
        }
      }
    }
    if (sentIdx > 0) {
      chunks.push(remaining.slice(0, sentIdx).trimEnd());
      remaining = remaining.slice(sentIdx).trimStart();
      continue;
    }

    // Word boundary (space or \n)
    let wordIdx = -1;
    for (let i = chunk.length - 1; i >= 0; i--) {
      if (chunk[i] === ' ' || chunk[i] === '\n') {
        wordIdx = i;
        break;
      }
    }
    if (wordIdx > 0) {
      chunks.push(remaining.slice(0, wordIdx).trimEnd());
      remaining = remaining.slice(wordIdx).trimStart();
      continue;
    }

    // Hard split at maxLen
    chunks.push(remaining.slice(0, maxLen));
    remaining = remaining.slice(maxLen);
  }

  if (remaining.trim() !== '') {
    chunks.push(remaining);
  }

  return chunks;
}

// Silence unused variable warning — metadataCallback is stored but only
// invoked by callers who register via onMetadata(). onMetadata is part of the
// interface contract even if Discord doesn't currently emit metadata events.
void (undefined as unknown as typeof DiscordChannel.prototype.onMetadata);
